x-pack/plugins/maps/public/classes/layers/layer_group/layer_group.tsx

```diff
@@ -0,0 +1,384 @@
+import { copyPersistentState } from '../../../reducers/copy_persistent_state';
+import {
+  Attribution,
+  CustomIcon,
+  LayerDescriptor,
+  LayerGroupDescriptor,
+  MapExtent,
+  StyleDescriptor,
+  StyleMetaDescriptor,
+} from '../../../../common/descriptor_types';
+import { ImmutableSourceProperty, ISource, SourceEditorArgs } from '../../sources/source';
+import { type DataRequestContext } from '../../../actions';
+import { getLayersExtent } from '../../../actions/get_layers_extent';
+import { ILayer, LayerIcon } from '../layer';
+import { IStyle } from '../../styles/style';
+import { LICENSED_FEATURES } from '../../../licensed_features';
+
+export function isLayerGroup(layer: ILayer) {
+  return layer instanceof LayerGroup;
+}
+
+export class LayerGroup implements ILayer {
+  protected readonly _descriptor: LayerGroupDescriptor;
+  private _children: ILayer[] = [];
+
+  static createDescriptor(options: Partial<LayerDescriptor>): LayerGroupDescriptor {
+    return {
+      ...options,
+      type: LAYER_TYPE.LAYER_GROUP,
+      id: typeof options.id === 'string' && options.id.length ? options.id : uuid(),
+      label:
+        typeof options.label === 'string' && options.label.length
+          ? options.label
+          : i18n.translate('xpack.maps.layerGroup.defaultName', {
+              defaultMessage: 'Layer group',
+            }),
+      sourceDescriptor: null,
+      visible: typeof options.visible === 'boolean' ? options.visible : true,
+    };
+  }
+
+  constructor({ layerDescriptor }: { layerDescriptor: LayerGroupDescriptor }) {
+    this._descriptor = LayerGroup.createDescriptor(layerDescriptor);
+  }
+
+  setChildren(children: ILayer[]) {
+    this._children = children;
+  }
+
+  getChildren(): ILayer[] {
+    return [...this._children];
+  }
+
+  async _asyncSomeChildren(methodName: string) {
+    const promises = this.getChildren().map(async (child) => {
+      // @ts-ignore
+      return (child[methodName] as () => Promise<boolean>)();
+    });
+    return ((await Promise.all(promises)) as boolean[]).some((result) => {
+      return result;
+    });
+  }
+
+  getDescriptor(): LayerGroupDescriptor {
+    return this._descriptor;
+  }
+
+  async cloneDescriptor(): Promise<LayerDescriptor[]> {
+    const clonedDescriptor = copyPersistentState(this._descriptor);
+    clonedDescriptor.id = uuid();
+    const displayName = await this.getDisplayName();
+    clonedDescriptor.label = `Clone of ${displayName}`;
+
+    const clonedChildrenDescriptors: LayerDescriptor[] = [];
+    await asyncForEach(this.getChildren(), async (childLayer) => {
+      (await childLayer.cloneDescriptor()).forEach((childLayerDescriptor) => {
+        if (childLayerDescriptor.parent === this.getId()) {
+          childLayerDescriptor.parent = clonedDescriptor.id;
+        }
+        clonedChildrenDescriptors.push(childLayerDescriptor);
+      });
+    });
+    return [...clonedChildrenDescriptors, clonedDescriptor];
```

Comment:
> I'm not sure why, but layer order in a cloned layer group is not guaranteed to be the same as the originating layer group.

The reason why order is not guaranteed is that `cloneDescriptor` is async. So when calling `cloneDescriptor` on all children, the order of return is not guaranteed and descriptors are added to array in returned order. Is this a big problem? Maybe the children array could be sorted after all async calls have returned?

Reply:
> Is this a big problem?

Not a critical issue, just unexpected. I have a hunch that cloning layers may not be very common, but I don't have telemetry to back up that statement. 

> So when calling cloneDescriptor on all children, the order of return is not guaranteed and descriptors are added to array in returned order. 

Could we built the `clonedChildrenDescriptors` array with `Promise.all` instead? I think that would return the correct order.

Reply:
Good idea using asyncMap instead of asyncForEach. That does preserve order. https://github.com/elastic/kibana/pull/142528/commits/1c1cfdb62609ae74e81563f9d96f5ef51a3895b0